File: M11AI3VB/ChatUI.vb
Imports Microsoft.Xna.Framework
Imports Microsoft.Xna.Framework.Graphics
Imports Microsoft.Xna.Framework.Input

Public NotInheritable Class ChatUI
    Implements IDisposable

    Private ReadOnly _font As SpriteFont
    Private ReadOnly _pixel As Texture2D
    Private ReadOnly _graphicsDevice As GraphicsDevice
    Private _userIcon As Texture2D
    Private _aiIcon As Texture2D
    Private _isDisposed As Boolean = False

    Private ReadOnly _messages As List(Of ChatMessage) = New List(Of ChatMessage)()
    Private _inputText As String = String.Empty
    Private _inputActive As Boolean = True
    Private _waitingForBotReply As Boolean = False
    Private _botReplyTimer As Single = 0F
    Private _pendingBotReply As String = String.Empty
    Private _previousKeyboardState As KeyboardState
    Private _previousMouseState As MouseState

    ' Scrolling variables
    Private _scrollOffset As Integer = 0
    Private _maxScroll As Integer = 0
    Private _scrollBarRect As Rectangle
    Private _isDraggingScrollBar As Boolean = False
    Private _lastMouseDragPosition As Vector2
    Private Const ChatBoxMargin As Integer = 10
    Private Const InputBoxHeight As Integer = 50
    Private Const SendButtonWidth As Integer = 80
    Private Const LineSpacing As Integer = 5
    Private Const MaxLineWidth As Integer = 500
    Private Const MessagePadding As Integer = 10
    Private Const BubblePadding As Integer = 15
    Private Const IconSize As Integer = 32
    Private Const ScrollBarWidth As Integer = 10
    Private Const ScrollBarMinHeight As Integer = 20

    Public Sub New(font As SpriteFont, graphicsDevice As GraphicsDevice)
        _font = font
        _graphicsDevice = graphicsDevice
        _pixel = New Texture2D(graphicsDevice, 1, 1)
        _pixel.SetData({Color.White})

        _previousKeyboardState = Keyboard.GetState()
        _previousMouseState = Mouse.GetState()

        AddMessage("Mono11", "Hello! I am Monoeleven AI. How can I help you?")
    End Sub

    Public Sub LoadIcons(userIcon As Texture2D, aiIcon As Texture2D)
        _userIcon = userIcon
        _aiIcon = aiIcon
    End Sub

    Public Sub Update(gameTime As GameTime)
        Dim keyboardState = Keyboard.GetState()
        Dim mouseState = Mouse.GetState()

        If _inputActive AndAlso Not _waitingForBotReply Then
            HandleTextInput(keyboardState)
            HandleMouseInput(mouseState)
        End If

        If _waitingForBotReply Then
            _botReplyTimer += CSng(gameTime.ElapsedGameTime.TotalSeconds)
            If _botReplyTimer >= 1.0F Then
                AddMessage("Mono11", _pendingBotReply)
                _waitingForBotReply = False
                _botReplyTimer = 0F
                _pendingBotReply = String.Empty
                _inputActive = True
            End If
        End If

        _previousKeyboardState = keyboardState
        _previousMouseState = mouseState
    End Sub

    Private Sub HandleTextInput(keyboardState As KeyboardState)
        For Each key In keyboardState.GetPressedKeys()
            If Not _previousKeyboardState.IsKeyDown(key) Then
                If key = Keys.Enter AndAlso _inputText.Length > 0 Then
                    SendMessage()
                    Return
                End If

                If key = Keys.Back AndAlso _inputText.Length > 0 Then
                    _inputText = _inputText.Substring(0, _inputText.Length - 1)
                    Continue For
                End If

                Dim character = GetCharacterFromKey(key, keyboardState)
                If Not Equals(character, Nothing) AndAlso _inputText.Length < 100 Then
                    _inputText &= character
                End If
            End If
        Next
    End Sub

    Private Shared Function GetCharacterFromKey(key As Keys, keyboardState As KeyboardState) As String
        Dim shiftPressed = keyboardState.IsKeyDown(Keys.LeftShift) OrElse keyboardState.IsKeyDown(Keys.RightShift)

        Select case key
            Case Keys.Space : Return " "
            Case Keys.A : Return If(shiftPressed, "A", "a")
            Case Keys.B : Return If(shiftPressed, "B", "b")
            Case Keys.C : Return If(shiftPressed, "C", "c")
            Case Keys.D : Return If(shiftPressed, "D", "d")
            Case Keys.E : Return If(shiftPressed, "E", "e")
            Case Keys.F : Return If(shiftPressed, "F", "f")
            Case Keys.G : Return If(shiftPressed, "G", "g")
            Case Keys.H : Return If(shiftPressed, "H", "h")
            Case Keys.I : Return If(shiftPressed, "I", "i")
            Case Keys.J : Return If(shiftPressed, "J", "j")
            Case Keys.K : Return If(shiftPressed, "K", "k")
            Case Keys.L : Return If(shiftPressed, "L", "l")
            Case Keys.M : Return If(shiftPressed, "M", "m")
            Case Keys.N : Return If(shiftPressed, "N", "n")
            Case Keys.O : Return If(shiftPressed, "O", "o")
            Case Keys.P : Return If(shiftPressed, "P", "p")
            Case Keys.Q : Return If(shiftPressed, "Q", "q")
            Case Keys.R : Return If(shiftPressed, "R", "r")
            Case Keys.S : Return If(shiftPressed, "S", "s")
            Case Keys.T : Return If(shiftPressed, "T", "t")
            Case Keys.U : Return If(shiftPressed, "U", "u")
            Case Keys.V : Return If(shiftPressed, "V", "v")
            Case Keys.W : Return If(shiftPressed, "W", "w")
            Case Keys.X : Return If(shiftPressed, "X", "x")
            Case Keys.Y : Return If(shiftPressed, "Y", "y")
            Case Keys.Z : Return If(shiftPressed, "Z", "z")
            Case Keys.D0 : Return If(shiftPressed, ")", "0")
            Case Keys.D1 : Return If(shiftPressed, "!", "1")
            Case Keys.D2 : Return If(shiftPressed, "@", "2")
            Case Keys.D3 : Return If(shiftPressed, "#", "3")
            Case Keys.D4 : Return If(shiftPressed, "$", "4")
            Case Keys.D5 : Return If(shiftPressed, "%", "5")
            Case Keys.D6 : Return If(shiftPressed, "^", "6")
            Case Keys.D7 : Return If(shiftPressed, "&", "7")
            Case Keys.D8 : Return If(shiftPressed, "*", "8")
            Case Keys.D9 : Return If(shiftPressed, "(", "9")
            Case Keys.OemComma : Return If(shiftPressed, "<", ",")
            Case Keys.OemPeriod : Return If(shiftPressed, ">", ".")
            Case Keys.OemQuestion : Return If(shiftPressed, "?", "/")
            Case Keys.OemSemicolon : Return If(shiftPressed, ":", ";")
            Case Keys.OemQuotes : Return If(shiftPressed, """", "'")
            Case Keys.OemTilde : Return If(shiftPressed, "~", "`")
            Case Else : Return Nothing
        End Select
    End Function

    Private Function WrapText(text As String, maxWidth As Single) As List(Of String)
        Dim words = text.Split(" "c)
        Dim lines As New List(Of String)()
        Dim currentLine As New System.Text.StringBuilder()
        Dim currentLineWidth As Single = 0F

        For Each word In words
            Dim wordWidth = _font.MeasureString(word).X
            Dim spaceWidth = If(currentLine.Length > 0, _font.MeasureString(" ").X, 0F)

            If currentLine.Length = 0 Then
                currentLine.Append(word)
                currentLineWidth = wordWidth
            ElseIf currentLineWidth + spaceWidth + wordWidth <= maxWidth Then
                currentLine.Append(" " & word)
                currentLineWidth += spaceWidth + wordWidth
            Else
                lines.Add(currentLine.ToString())
                currentLine.Clear()
                currentLine.Append(word)
                currentLineWidth = wordWidth
            End If
        Next word
        If currentLine.Length > 0 Then lines.Add(currentLine.ToString())
        If lines.Count = 0 Then lines.Add(text)
        Return lines
    End Function

    Private Sub HandleMouseInput(mouseState As MouseState)
        Dim inputBoxRect As New Rectangle(
            ChatBoxMargin,
            _graphicsDevice.Viewport.Height - InputBoxHeight - ChatBoxMargin,
            _graphicsDevice.Viewport.Width - ChatBoxMargin * 2,
            InputBoxHeight
        )
        Dim sendButtonRect As New Rectangle(
            inputBoxRect.Right - SendButtonWidth - 5, inputBoxRect.Y + 5,
            SendButtonWidth, inputBoxRect.Height - 10
        )

        ' Handle send button click
        If mouseState.LeftButton = ButtonState.Pressed AndAlso
            _previousMouseState.LeftButton = ButtonState.Released AndAlso
            sendButtonRect.Contains(mouseState.Position) AndAlso
            _inputText.Length > 0 Then SendMessage()

        ' Handle scrollbar interaction
        Dim chatBoxRect As New Rectangle(
            ChatBoxMargin,
            ChatBoxMargin,
            _graphicsDevice.Viewport.Width - ChatBoxMargin * 2,
            _graphicsDevice.Viewport.Height - InputBoxHeight - ChatBoxMargin * 3
        )

        If mouseState.LeftButton = ButtonState.Pressed AndAlso
            _previousMouseState.LeftButton = ButtonState.Released AndAlso
            _scrollBarRect.Contains(mouseState.Position) AndAlso _maxScroll > 0 Then
            ' Start dragging scrollbar
            _isDraggingScrollBar = True
            _lastMouseDragPosition = New Vector2(mouseState.X, mouseState.Y)
        ElseIf mouseState.LeftButton = ButtonState.Released Then
            ' Stop dragging scrollbar
            _isDraggingScrollBar = False
        ElseIf _isDraggingScrollBar Then
            ' Continue dragging scrollbar
            Dim deltaY = mouseState.Y - _lastMouseDragPosition.Y
            _lastMouseDragPosition = New Vector2(mouseState.X, mouseState.Y)

            ' Calculate how much to scroll based on drag distance
            Dim scrollRatio = CSng(chatBoxRect.Height - 20 - _scrollBarRect.Height) / _maxScroll
            _scrollOffset = MathHelper.Clamp(_scrollOffset + CInt(deltaY / scrollRatio), 0, _maxScroll)
        End If

        ' Handle mouse wheel scrolling
        Dim mouseWheelDelta = mouseState.ScrollWheelValue - _previousMouseState.ScrollWheelValue
        If mouseWheelDelta <> 0 AndAlso chatBoxRect.Contains(New Point(mouseState.X, mouseState.Y)) Then
            Dim scrollAmount = -Math.Sign(mouseWheelDelta) * 30 ' Adjust sensitivity
            _scrollOffset = MathHelper.Clamp(_scrollOffset + scrollAmount, 0, _maxScroll)
        End If
    End Sub

    Private Sub SendMessage()
        If String.IsNullOrWhiteSpace(_inputText) Then Exit Sub

        AddMessage("User", _inputText)
        _pendingBotReply = ChatbotReply(_inputText)
        _waitingForBotReply = True
        _inputActive = False
        _inputText = String.Empty
    End Sub

    Private Sub AddMessage(sender As String, text As String)
        _messages.Add(New ChatMessage With {.Sender = sender, .Text = text})
    End Sub

    Public Sub Draw(spriteBatch As SpriteBatch, screenWidth As Integer, screenHeight As Integer, Optional positionOffset As Point = DirectCast(Nothing, Point))
        Dim chatBoxRect As New Rectangle(
            ChatBoxMargin + positionOffset.X, 
            ChatBoxMargin + positionOffset.Y, 
            screenWidth - ChatBoxMargin * 2, 
            screenHeight - InputBoxHeight - ChatBoxMargin * 3)
        Dim inputBoxRect As New Rectangle(
            ChatBoxMargin + positionOffset.X, 
            positionOffset.Y + screenHeight - InputBoxHeight - ChatBoxMargin, 
            screenWidth - ChatBoxMargin * 2, InputBoxHeight)
        Dim sendButtonRect As New Rectangle(inputBoxRect.Right - SendButtonWidth - 5, inputBoxRect.Y + 5, SendButtonWidth, inputBoxRect.Height - 10)

        Me.DrawChatBox(spriteBatch, chatBoxRect)
        Me.DrawInputBox(spriteBatch, inputBoxRect, sendButtonRect)
    End Sub

    Private Sub DrawChatBox(spriteBatch As SpriteBatch, rect As Rectangle)
        spriteBatch.Draw(_pixel, rect, Color.WhiteSmoke)
        spriteBatch.Draw(_pixel, New Rectangle(rect.X, rect.Y, rect.Width, 2), Color.Gray)

        Dim yPos = rect.Y + 10 - _scrollOffset
        For Each message In _messages
            Dim wrappedLines = WrapText(message.Text, ChatUI.MaxLineWidth)
            Dim maxLineWidth As Single = 0F
            For Each line In wrappedLines
                Dim lineWidth = _font.MeasureString(line).X
                If lineWidth > maxLineWidth Then
                    maxLineWidth = lineWidth
                End If
            Next

            Dim totalTextHeight = wrappedLines.Count * _font.LineSpacing + (wrappedLines.Count - 1) * LineSpacing
            Dim bubbleHeight = Math.Max(totalTextHeight + BubblePadding * 2, IconSize + BubblePadding * 2)
            Dim bubbleWidth = maxLineWidth + BubblePadding * 2 + IconSize + MessagePadding

            Dim isUserMessage = message.Sender Is "User"
            Dim bubbleX = If(isUserMessage, rect.Right - bubbleWidth - 20, rect.X + 10)

            Dim bubbleRect = New Rectangle(CInt(MathF.Round(bubbleX)), yPos, CInt(bubbleWidth), CInt(bubbleHeight))
            Dim bubbleColor = If(isUserMessage, New Color(220, 248, 198), New Color(230, 230, 250))
            spriteBatch.Draw(_pixel, bubbleRect, bubbleColor)

            Dim iconTexture = If(isUserMessage, _userIcon, _aiIcon)
            If iconTexture IsNot Nothing Then
                Dim iconRect = New Rectangle(If(isUserMessage, bubbleRect.Right - IconSize - 5, bubbleRect.X + 5), CInt(bubbleRect.Y + (bubbleHeight - IconSize) / 2), IconSize, IconSize)
                spriteBatch.Draw(iconTexture, iconRect, Color.White)
            End If

            Dim textStartY = bubbleRect.Y + (bubbleHeight - totalTextHeight) / 2
            For i = 0 To wrappedLines.Count - 1
                Dim lineText = wrappedLines(i)
                Dim textSize = _font.MeasureString(lineText)
                Dim textPosition = New Vector2(If(isUserMessage, bubbleRect.Right - textSize.X - IconSize - MessagePadding, bubbleRect.X + IconSize + MessagePadding), textStartY + i * (_font.LineSpacing + LineSpacing))
                spriteBatch.DrawString(_font, lineText, textPosition, Color.Black)
            Next

            yPos += CInt(bubbleHeight) + MessagePadding
        Next

        Dim totalMessagesHeight = 0F
        For Each message In _messages
            Dim wrappedLines = WrapText(message.Text, ChatUI.MaxLineWidth)
            Dim maxLineWidth As Single = 0F
            For Each line In wrappedLines
                Dim lineWidth = _font.MeasureString(line).X
                If lineWidth > maxLineWidth Then
                    maxLineWidth = lineWidth
                End If
            Next

            Dim totalTextHeight = wrappedLines.Count * _font.LineSpacing + (wrappedLines.Count - 1) * LineSpacing
            Dim bubbleHeight = Math.Max(totalTextHeight + BubblePadding * 2, IconSize + BubblePadding * 2)
            totalMessagesHeight += bubbleHeight + MessagePadding
        Next

        _maxScroll = Math.Max(0, CInt(totalMessagesHeight - rect.Height + 20))

        Dim scrollBarHeight = Math.Max(ScrollBarMinHeight, CInt(rect.Height / Math.Max(totalMessagesHeight, rect.Height) * rect.Height))
        Dim scrollBarY = rect.Y + 10 + CInt(_scrollOffset / CSng(Math.Max(_maxScroll, 1)) * (rect.Height - 10 - scrollBarHeight))

        _scrollBarRect = New Rectangle(rect.Right - ScrollBarWidth - 5, scrollBarY, ScrollBarWidth, scrollBarHeight)

        spriteBatch.Draw(_pixel, New Rectangle(_scrollBarRect.X - 2, rect.Y + 10, ScrollBarWidth + 4, rect.Height - 20), Color.LightGray)

        spriteBatch.Draw(_pixel, _scrollBarRect, If(_isDraggingScrollBar, Color.DarkGray, Color.Gray))
    End Sub

    Private Sub DrawInputBox(spriteBatch As SpriteBatch, inputRect As Rectangle, sendButtonRect As Rectangle)
        spriteBatch.Draw(_pixel, inputRect, Color.White)
        spriteBatch.Draw(_pixel, New Rectangle(inputRect.X, inputRect.Y, inputRect.Width, 2), Color.Gray)

        Dim inputTextPos = New Vector2(inputRect.X + 10, inputRect.Y + 15)
        spriteBatch.DrawString(_font, _inputText & If(_inputActive, "|", ""), inputTextPos, Color.Black)

        spriteBatch.Draw(_pixel, sendButtonRect, Color.LightBlue)
        spriteBatch.Draw(_pixel, New Rectangle(sendButtonRect.X, sendButtonRect.Y, sendButtonRect.Width, 2), Color.Blue)

        Dim buttonText = "Send"
        Dim buttonTextSize = _font.MeasureString(buttonText)
        Dim buttonTextPos = New Vector2(sendButtonRect.X + (sendButtonRect.Width - buttonTextSize.X) / 2, sendButtonRect.Y + (sendButtonRect.Height - buttonTextSize.Y) / 2)
        spriteBatch.DrawString(_font, buttonText, buttonTextPos, Color.Black)
    End Sub

    Private Class ChatMessage
        Public Property Sender As String = String.Empty
        Public Property Text As String = String.Empty
    End Class

    Public Sub Dispose() Implements IDisposable.Dispose
        Dispose(True)
        GC.SuppressFinalize(Me)
    End Sub

    Private Sub Dispose(disposing As Boolean)
        If Not _isDisposed Then
            If disposing Then
                _pixel.Dispose()
            End If
            _isDisposed = True
        End If
    End Sub
End Class

File: M11AI3CSharp/ChatUI.cs
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;

namespace M11AI3CSharp;

public sealed class ChatUI : IDisposable
{
    private readonly SpriteFont _font;
    private readonly Texture2D _pixel;
    private readonly GraphicsDevice _graphicsDevice;
    private Texture2D? _userIcon;
    private Texture2D? _aiIcon;
    private bool _isDisposed = false;

    private readonly List<ChatMessage> _messages = [];
    private string _inputText = string.Empty;
    private bool _inputActive = true;
    private bool _waitingForBotReply = false;
    private float _botReplyTimer = 0f;
    private string _pendingBotReply = string.Empty;
    private KeyboardState _previousKeyboardState;
    private MouseState _previousMouseState;
    
    // Scrolling variables
    private int _scrollOffset = 0;
    private int _maxScroll = 0;
    private Rectangle _scrollBarRect;
    private bool _isDraggingScrollBar = false;
    private Vector2 _lastMouseDragPosition;

    private const int ChatBoxMargin = 10;
    private const int InputBoxHeight = 50;
    private const int SendButtonWidth = 80;
    private const int LineSpacing = 5;
    private const int MessagePadding = 10;
    private const int BubblePadding = 15;
    private const int IconSize = 32;
    private const int ScrollBarWidth = 10;
    private const int ScrollBarMinHeight = 20;
    private const int MaxLineWidth = 500;

    public ChatUI(SpriteFont font, GraphicsDevice graphicsDevice)
    {
        _font = font;
        _graphicsDevice = graphicsDevice;
        _pixel = new Texture2D(graphicsDevice, 1, 1);
        _pixel.SetData([Color.White]);

        _previousKeyboardState = Keyboard.GetState();
        _previousMouseState = Mouse.GetState();

        AddMessage("Mono11", "Hello! I am Monoeleven AI. How can I help you?");
    }
    
    public void LoadIcons(Texture2D userIcon, Texture2D aiIcon)
    {
        _userIcon = userIcon;
        _aiIcon = aiIcon;
    }

    public void Update(GameTime gameTime)
    {
        var keyboardState = Keyboard.GetState();
        var mouseState = Mouse.GetState();

        if (_inputActive && !_waitingForBotReply)
        {
            HandleTextInput(keyboardState);
            HandleMouseInput(mouseState);
        }

        if (_waitingForBotReply)
        {
            _botReplyTimer += (float)gameTime.ElapsedGameTime.TotalSeconds;
            if (_botReplyTimer >= 1.0f)
            {
                AddMessage("Mono11", _pendingBotReply);
                _waitingForBotReply = false;
                _botReplyTimer = 0f;
                _pendingBotReply = string.Empty;
                _inputActive = true;
            }
        }

        _previousKeyboardState = keyboardState;
        _previousMouseState = mouseState;
    }

    private void HandleTextInput(KeyboardState keyboardState)
    {
        var keys = keyboardState.GetPressedKeys();

        foreach (var key in keys)
        {
            if (!_previousKeyboardState.IsKeyDown(key))
            {
                if (key == Keys.Enter && _inputText.Length > 0)
                {
                    SendMessage();
                    return;
                }

                if (key == Keys.Back && _inputText.Length > 0)
                {
                    _inputText = _inputText.Substring(0, _inputText.Length - 1);
                    continue;
                }

                var character = GetCharacterFromKey(key, keyboardState);
                if (character != null && _inputText.Length < 100)
                {
                    _inputText += character;
                }
            }
        }
    }

    private static string? GetCharacterFromKey(Keys key, KeyboardState keyboardState)
    {
        var shiftPressed = keyboardState.IsKeyDown(Keys.LeftShift) || keyboardState.IsKeyDown(Keys.RightShift);

        return key switch
        {
            Keys.Space => " ",
            Keys.A => shiftPressed ? "A" : "a",
            Keys.B => shiftPressed ? "B" : "b",
            Keys.C => shiftPressed ? "C" : "c",
            Keys.D => shiftPressed ? "D" : "d",
            Keys.E => shiftPressed ? "E" : "e",
            Keys.F => shiftPressed ? "F" : "f",
            Keys.G => shiftPressed ? "G" : "g",
            Keys.H => shiftPressed ? "H" : "h",
            Keys.I => shiftPressed ? "I" : "i",
            Keys.J => shiftPressed ? "J" : "j",
            Keys.K => shiftPressed ? "K" : "k",
            Keys.L => shiftPressed ? "L" : "l",
            Keys.M => shiftPressed ? "M" : "m",
            Keys.N => shiftPressed ? "N" : "n",
            Keys.O => shiftPressed ? "O" : "o",
            Keys.P => shiftPressed ? "P" : "p",
            Keys.Q => shiftPressed ? "Q" : "q",
            Keys.R => shiftPressed ? "R" : "r",
            Keys.S => shiftPressed ? "S" : "s",
            Keys.T => shiftPressed ? "T" : "t",
            Keys.U => shiftPressed ? "U" : "u",
            Keys.V => shiftPressed ? "V" : "v",
            Keys.W => shiftPressed ? "W" : "w",
            Keys.X => shiftPressed ? "X" : "x",
            Keys.Y => shiftPressed ? "Y" : "y",
            Keys.Z => shiftPressed ? "Z" : "z",
            Keys.D0 => shiftPressed ? ")" : "0",
            Keys.D1 => shiftPressed ? "!" : "1",
            Keys.D2 => shiftPressed ? "@" : "2",
            Keys.D3 => shiftPressed ? "#" : "3",
            Keys.D4 => shiftPressed ? "$" : "4",
            Keys.D5 => shiftPressed ? "%" : "5",
            Keys.D6 => shiftPressed ? "^" : "6",
            Keys.D7 => shiftPressed ? "&" : "7",
            Keys.D8 => shiftPressed ? "*" : "8",
            Keys.D9 => shiftPressed ? "(" : "9",
            Keys.OemComma => shiftPressed ? "<" : ",",
            Keys.OemPeriod => shiftPressed ? ">" : ".",
            Keys.OemQuestion => shiftPressed ? "?" : "/",
            Keys.OemSemicolon => shiftPressed ? ":" : ";",
            Keys.OemQuotes => shiftPressed ? "\"" : "'",
            Keys.OemTilde => shiftPressed ? "~" : "`",
            _ => null
        };
    }

    private void HandleMouseInput(MouseState mouseState)
    {
        var inputBoxRect = new Rectangle(
            ChatBoxMargin,
            _graphicsDevice.Viewport.Height - InputBoxHeight - ChatBoxMargin,
            _graphicsDevice.Viewport.Width - ChatBoxMargin * 2,
            InputBoxHeight
        );

        var sendButtonRect = new Rectangle(
            inputBoxRect.Right - SendButtonWidth - 5,
            inputBoxRect.Y + 5,
            SendButtonWidth,
            inputBoxRect.Height - 10
        );

        if (mouseState.LeftButton == ButtonState.Pressed && 
            _previousMouseState.LeftButton == ButtonState.Released && 
            sendButtonRect.Contains(mouseState.Position) && 
            _inputText.Length > 0)
        {
            SendMessage();
        }
        
        var chatBoxRect = new Rectangle(
            ChatBoxMargin,
            ChatBoxMargin,
            _graphicsDevice.Viewport.Width - ChatBoxMargin * 2,
            _graphicsDevice.Viewport.Height - InputBoxHeight - ChatBoxMargin * 3
        );
        
        if (mouseState.LeftButton == ButtonState.Pressed && 
            _previousMouseState.LeftButton == ButtonState.Released &&
            _scrollBarRect.Contains(mouseState.Position) && 
            _maxScroll > 0)
        {
            _isDraggingScrollBar = true;
            _lastMouseDragPosition = new Vector2(mouseState.X, mouseState.Y);
        }
        else if (mouseState.LeftButton == ButtonState.Released)
        {
            _isDraggingScrollBar = false;
        }
        else if (_isDraggingScrollBar)
        {
            var deltaY = mouseState.Y - _lastMouseDragPosition.Y;
            _lastMouseDragPosition = new Vector2(mouseState.X, mouseState.Y);
            
            var scrollRatio = (float)(chatBoxRect.Height - 20 - _scrollBarRect.Height) / _maxScroll;
            _scrollOffset = MathHelper.Clamp(_scrollOffset + (int)(deltaY / scrollRatio), 0, _maxScroll);
        }
        
        var mouseWheelDelta = mouseState.ScrollWheelValue - _previousMouseState.ScrollWheelValue;
        if (mouseWheelDelta != 0 && chatBoxRect.Contains(new Point(mouseState.X, mouseState.Y)))
        {
            var scrollAmount = -Math.Sign(mouseWheelDelta) * 30;
            _scrollOffset = MathHelper.Clamp(_scrollOffset + scrollAmount, 0, _maxScroll);
        }
    }

    private List<string> WrapText(string text, float maxWidth)
    {
        var words = text.Split(' ');
        var lines = new List<string>();
        var currentLine = new StringBuilder();
        var currentLineWidth = 0f;

        foreach (var word in words)
        {
            var wordWidth = _font.MeasureString(word).X;
            var spaceWidth = currentLine.Length > 0 ? _font.MeasureString(" ").X : 0f;

            if (currentLine.Length == 0)
            {
                currentLine.Append(word);
                currentLineWidth = wordWidth;
            }
            else if (currentLineWidth + spaceWidth + wordWidth <= maxWidth)
            {
                currentLine.Append(" " + word);
                currentLineWidth += spaceWidth + wordWidth;
            }
            else
            {
                lines.Add(currentLine.ToString());
                currentLine.Clear();
                currentLine.Append(word);
                currentLineWidth = wordWidth;
            }
        }

        if (currentLine.Length > 0)
        {
            lines.Add(currentLine.ToString());
        }

        if (lines.Count == 0)
        {
            lines.Add(text);
        }

        return lines;
    }

    private void SendMessage()
    {
        if (string.IsNullOrWhiteSpace(_inputText)) return;

        AddMessage("User", _inputText);

        var botReply = ChatBotLogic.GetChatbotReply(_inputText);
        _pendingBotReply = botReply;
        _waitingForBotReply = true;
        _inputActive = false;
        _inputText = string.Empty;
    }

    private void AddMessage(string sender, string text)
    {
        _messages.Add(new ChatMessage { Sender = sender, Text = text });
    }

    public void Draw(SpriteBatch spriteBatch, int screenWidth, int screenHeight, Point positionOffset = default)
    {
        var chatBoxRect = new Rectangle(
            ChatBoxMargin + positionOffset.X,
            ChatBoxMargin + positionOffset.Y,
            screenWidth - ChatBoxMargin * 2,
            screenHeight - InputBoxHeight - ChatBoxMargin * 3
        );

        var inputBoxRect = new Rectangle(
            ChatBoxMargin + positionOffset.X,
            positionOffset.Y + screenHeight - InputBoxHeight - ChatBoxMargin,
            screenWidth - ChatBoxMargin * 2,
            InputBoxHeight
        );

        var sendButtonRect = new Rectangle(
            inputBoxRect.Right - SendButtonWidth - 5,
            inputBoxRect.Y + 5,
            SendButtonWidth,
            inputBoxRect.Height - 10
        );

        DrawChatBox(spriteBatch, chatBoxRect);
        DrawInputBox(spriteBatch, inputBoxRect, sendButtonRect);
    }

    private void DrawChatBox(SpriteBatch spriteBatch, Rectangle rect)
    {
        spriteBatch.Draw(_pixel, rect, Color.WhiteSmoke);
        spriteBatch.Draw(_pixel, new Rectangle(rect.X, rect.Y, rect.Width, 2), Color.Gray);

        var yPos = rect.Y + 10 - _scrollOffset;
        foreach (var message in _messages)
        {
            var wrappedLines = WrapText(message.Text, MaxLineWidth);
            var maxLineWidth = 0f;
            foreach (var line in wrappedLines)
            {
                var lineWidth = _font.MeasureString(line).X;
                if (lineWidth > maxLineWidth)
                {
                    maxLineWidth = lineWidth;
                }
            }

            var totalTextHeight = wrappedLines.Count * _font.LineSpacing + (wrappedLines.Count - 1) * LineSpacing;
            var bubbleHeight = Math.Max(totalTextHeight + BubblePadding * 2, IconSize + BubblePadding * 2);
            var bubbleWidth = maxLineWidth + BubblePadding * 2 + IconSize + MessagePadding;

            var isUserMessage = message.Sender == "User";
            var bubbleX = isUserMessage ? rect.Right - bubbleWidth - 20 : rect.X + 10;

            var bubbleRect = new Rectangle((int)MathF.Round(bubbleX), yPos, (int)bubbleWidth, (int)bubbleHeight);
            var bubbleColor = isUserMessage ? new Color(220, 248, 198) : new Color(230, 230, 250);
            spriteBatch.Draw(_pixel, bubbleRect, bubbleColor);

            var iconTexture = isUserMessage ? _userIcon : _aiIcon;
            if (iconTexture != null)
            {
                var iconRect = new Rectangle(
                    isUserMessage ? bubbleRect.Right - IconSize - 5 : bubbleRect.X + 5,
                    (int)(bubbleRect.Y + (bubbleHeight - IconSize) / 2),
                    IconSize,
                    IconSize
                );
                spriteBatch.Draw(iconTexture, iconRect, Color.White);
            }

            var textStartY = bubbleRect.Y + (bubbleHeight - totalTextHeight) / 2;
            for (int i = 0; i < wrappedLines.Count; i++)
            {
                var lineText = wrappedLines[i];
                var textSize = _font.MeasureString(lineText);
                var textPosition = new Vector2(
                    isUserMessage ? bubbleRect.Right - textSize.X - IconSize - MessagePadding : bubbleRect.X + IconSize + MessagePadding,
                    textStartY + i * (_font.LineSpacing + LineSpacing)
                );
                spriteBatch.DrawString(_font, lineText, textPosition, Color.Black);
            }

            yPos += (int)bubbleHeight + MessagePadding;
        }

        var totalMessagesHeight = 0f;
        foreach (var message in _messages)
        {
            var wrappedLines = WrapText(message.Text, MaxLineWidth);
            var maxLineWidth = 0f;
            foreach (var line in wrappedLines)
            {
                var lineWidth = _font.MeasureString(line).X;
                if (lineWidth > maxLineWidth)
                {
                    maxLineWidth = lineWidth;
                }
            }

            var totalTextHeight = wrappedLines.Count * _font.LineSpacing + (wrappedLines.Count - 1) * LineSpacing;
            var bubbleHeight = Math.Max(totalTextHeight + BubblePadding * 2, IconSize + BubblePadding * 2);
            totalMessagesHeight += bubbleHeight + MessagePadding;
        }

        _maxScroll = Math.Max(0, (int)(totalMessagesHeight - rect.Height + 20));

        var scrollBarHeight = Math.Max(
            ScrollBarMinHeight,
            (int)(rect.Height / Math.Max(totalMessagesHeight, rect.Height) * rect.Height)
        );
        var scrollBarY = rect.Y + 10 + (int)(
            _scrollOffset / (float)Math.Max(_maxScroll, 1) * (rect.Height - 10 - scrollBarHeight)
        );

        _scrollBarRect = new Rectangle(
            rect.Right - ScrollBarWidth - 5,
            scrollBarY,
            ScrollBarWidth,
            scrollBarHeight
        );

        spriteBatch.Draw(_pixel, new Rectangle(_scrollBarRect.X - 2, rect.Y + 10, ScrollBarWidth + 4, rect.Height - 20), Color.LightGray);
        spriteBatch.Draw(_pixel, _scrollBarRect, _isDraggingScrollBar ? Color.DarkGray : Color.Gray);
    }

    private void DrawInputBox(SpriteBatch spriteBatch, Rectangle inputRect, Rectangle sendButtonRect)
    {
        spriteBatch.Draw(_pixel, inputRect, Color.White);
        spriteBatch.Draw(_pixel, new Rectangle(inputRect.X, inputRect.Y, inputRect.Width, 2), Color.Gray);

        var inputTextPos = new Vector2(inputRect.X + 10, inputRect.Y + 15);
        spriteBatch.DrawString(_font, _inputText + (_inputActive ? "|" : ""), inputTextPos, Color.Black);

        spriteBatch.Draw(_pixel, sendButtonRect, Color.LightBlue);
        spriteBatch.Draw(_pixel, new Rectangle(sendButtonRect.X, sendButtonRect.Y, sendButtonRect.Width, 2), Color.Blue);

        var buttonText = "Send";
        var buttonTextSize = _font.MeasureString(buttonText);
        var buttonTextPos = new Vector2(
            sendButtonRect.X + (sendButtonRect.Width - buttonTextSize.X) / 2,
            sendButtonRect.Y + (sendButtonRect.Height - buttonTextSize.Y) / 2
        );
        spriteBatch.DrawString(_font, buttonText, buttonTextPos, Color.Black);
    }

    private class ChatMessage
    {
        public string Sender { get; set; } = string.Empty;
        public string Text { get; set; } = string.Empty;
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    private void Dispose(bool disposing)
    {
        if (!_isDisposed)
        {
            if (disposing)
            {
                _pixel.Dispose();
            }
            _isDisposed = true;
        }
    }
}